planner/core/common_plans.go

```diff
@@ -282,10 +282,15 @@ func (e *Execute) getPhysicalPlan(ctx context.Context, sctx sessionctx.Context,
 	if prepared.UseCache {
 		cacheKey = NewPSTMTPlanCacheKey(sctx.GetSessionVars(), e.ExecID, prepared.SchemaVersion)
 	}
-	tps := make([]*types.FieldType, len(e.UsingVars))
+	tps := make([]*types.FieldType, 0)
 	for i, param := range e.UsingVars {
-		name := param.(*expression.ScalarFunction).GetArgs()[0].String()
-		tps[i] = sctx.GetSessionVars().UserVarTypes[name]
+		fun, ok := param.(*expression.ScalarFunction)
+		if !ok { // GetVar folded by scalar_function.go:newFunctionImpl()
```

Comment:
We can't skip the `UserVarType`, because plan cache uses it to distinguish different plans of different variable types.

Reply:
so...  When it is `Constant`, I should assgin the `c.GetType()` to typs[i]?

```go
for i, param := range e.UsingVars {
	fun, ok := param.(*expression.ScalarFunction)
	if ok {
		name := fun.GetArgs()[0].String()
		tps[i] = sctx.GetSessionVars().UserVarTypes[name]
	} else {
		// GetVar folded by scalar_function.go:newFunctionImpl()
		c := param.(*expression.Constant)
		tps[i] = c.GetType()
	}
	if tps[i] == nil {
		tps[i] = types.NewFieldType(mysql.TypeNull)
	}
}
```

Reply:
Yes.